src/graph/graphcontext.c

```diff
@@ -0,0 +1,184 @@
+  RedisModule_CloseKey(key);
+
+  gc->ctx = ctx;
+  gc->g = g;
+  gc->node_allstore = LabelStore_Get(ctx, STORE_NODE, graph_name, NULL);
+  gc->relation_allstore = LabelStore_Get(ctx, STORE_EDGE, graph_name, NULL);
+
+  // Retrieve all label stores?
+  // Can iterate over all label strings or serialize them within GraphContext value
+
+  return gc;
+}
+
+// TODO next two functions are nearly identical; maybe consolidate
+LabelStore* GraphContext_AddNode(GraphContext *gc, const char *label) {
+  if(gc->node_count == gc->node_cap) {
+    gc->node_cap += 4;
+    gc->node_stores = realloc(gc->node_stores, gc->node_cap * sizeof(LabelStore*));
+    memset(gc->node_stores + gc->node_count, 0, 4 * sizeof(LabelStore*));
+  }
+  // Graph_AddLabel(gc->g);
+  LabelStore *store = LabelStore_New(gc->ctx, STORE_NODE, gc->graph_name, label, gc->node_count);
+  gc->node_stores[gc->node_count] = store;
+  gc->node_count++;
+
+  return store;
+}
+
+LabelStore* GraphContext_AddRelation(GraphContext *gc, const char *label) {
+  if(gc->relation_count == gc->relation_cap) {
+    gc->relation_cap += 4;
+    gc->relation_stores = realloc(gc->relation_stores, gc->relation_cap * sizeof(LabelStore*));
+    memset(gc->relation_stores + gc->relation_count, 0, 4 * sizeof(LabelStore*));
+  }
+
+  LabelStore *store = LabelStore_New(gc->ctx, STORE_EDGE, gc->graph_name, label, gc->relation_count);
+  gc->relation_stores[gc->relation_count] = store;
+  gc->relation_count++;
+
+  return store;
+}
+
+int GraphContext_GetLabelID(const GraphContext *gc, const char *label, LabelStoreType t) {
+  LabelStore **labels;
+  size_t count;
+  if (t == STORE_NODE) {
+    labels = gc->node_stores;
+    count = gc->node_count;
+  } else {
+    labels = gc->relation_stores;
+    count = gc->relation_count;
+  }
+
+  for (int i = 0; i < count; i ++) {
+    if (labels[i] && !strcmp(label, labels[i]->label)) return i;
+  }
+  return GRAPH_NO_LABEL;
+}
+
+Index* GraphContext_GetIndex(GraphContext *gc, const char *label, const char *property) {
+  Index *idx;
+  for (int i = 0; i < gc->index_count; i ++) {
+    idx = gc->indices[i];
+    if (!strcmp(label, idx->label) && !strcmp(property, idx->property)) return idx;
+  }
+  return NULL;
+}
+
+bool GraphContext_HasIndices(GraphContext *gc) {
+  return gc->index_count > 0;
+}
+
+void GraphContext_AddIndex(GraphContext *gc, Index *idx) {
+  gc->index_count++;
+  if (!gc->indices) {
+    gc->indices = malloc(gc->index_count * sizeof(Index*));
+  } else {
+    gc->indices = realloc(gc->indices, gc->index_count * sizeof(Index*));
+  }
+  gc->indices[gc->index_count - 1] = idx;
+}
+
+LabelStore* GraphContext_AllStore(const GraphContext *gc, LabelStoreType t) {
+  if (t == STORE_NODE) return gc->node_allstore;
+  return gc->relation_allstore;
+}
+
+LabelStore* GraphContext_GetNodeStore(const GraphContext *gc, const char *label) {
+  // Check cached stores
+  int id = GraphContext_GetLabelID(gc, label, STORE_NODE);
```

Comment:
When the GraphContext is loaded from rdb, it won't have any stores associated, nor can it safely look them up (since we have no RedisModuleCtx and no guarantee that the stores themselves have been loaded yet).

I'm thinking now that this caching and key separation is overcomplicated, though, and I'm going to try just serializing stores as part of the GraphContext.